packages/muze-utils/src/store.js

```diff
@@ -0,0 +1,578 @@
+/**
+ *  Common store class
+ *
+ * @class Store
+ */
+export class Store {
+    /**
+     * Creates an instance of Store.
+     * @param {Object} config The object to create the state store with.
+     * @memberof Store
+     */
+    constructor (config) {
+        // create reactive model
+        this.model = Model.create(config);
+        this._listeners = new Map();
+        this._registeredListeners = {};
+        this._contextMap = {};
+        this._commits = {};
+        this._savedCommits = {};
+        this._queuedProps = {};
+        this._propListenerMap = {};
+        this._locked = false;
+    }
+
+    lockModel () {
+        this.model.lock();
+        this._locked = true;
+        return this;
+    }
+
+    unlockModel () {
+        this._locked = false;
+        this.model.unlock();
+        return this;
+    }
+
+    /**
+     * This method returns a plain JSON object
+     * with all the fields in the state store.
+     *
+     * @return {Object} Serialized representation of state store.
+     * @memberof Store
+     */
+    serialize () {
+        return this.model.serialize();
+    }
+
+    lockCommits (props) {
+        props.forEach((prop) => {
+            this._commits[prop] = {
+                locked: true,
+                queue: []
+            };
+        });
+        return this;
+    }
+
+    unlockCommits (props) {
+        this.lockModel();
+        const commitsObj = this._commits;
+        props.forEach((prop) => {
+            commitsObj[prop].locked = false;
+            const queuedProps = {};
+            const { queue } = commitsObj[prop];
+            queue.forEach((params) => {
+                const [propName, value, namespace] = params;
+                if (namespace) {
+                    const propObj = initProp(queuedProps, [propName, namespace]);
+                    Object.assign(propObj, value);
+                }
+            });
+            queue.forEach((params) => {
+                const [propName, value, namespace] = params;
+                if (propName in queuedProps) {
+                    this.commit(propName, queuedProps[propName][namespace], namespace);
+                } else {
+                    this.commit(propName, value, namespace);
+                }
+            });
+            delete commitsObj[prop];
+        });
+        this.unlockModel();
+        return this;
+    }
+
+    addSubNamespace (sns, namespace, context) {
+        // Get all the listeners registered by the component
+        const listeners = this._registeredListeners[namespace];
+        const propListenerMap = this._propListenerMap;
+
+        initProp(this._contextMap, [namespace]);
+        this._contextMap[namespace][sns] = context;
+        for (const key in listeners) {
+            const obj = listeners[key];
+            const propObj = propListenerMap[key];
+            const propFns = types.reduce((acc, type) => {
+                const val = defaultValue(getObjProp(propObj, type, 'fns'), 0);
```

Comment:
Preferably declare those const variables first that do not use any processes to be initialized.